src/Simulation/Simulators/OpenSystemsSimulator/DataModel/ArrayConverter.cs

```diff
@@ -3,249 +3,245 @@
-    ///       components of each element.
-    ///     <item><c>"data"</c>: Lists the elements of the array, with the
-    ///       right-most dimension varying the fastest (defined by <c>ndarray</c>
-    ///       as "logical order", and by NumPy as "C-ordered").
-    ///   </list>
-    /// </remarks>
-    public class ComplexArrayConverter : JsonConverter<NDArray>
+    public override NDArray Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
     {
-        public override NDArray Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        // By the end, we need to have read "v", "dim", and "data".
+        int? version = null;
+        List<int>? dims = null;
+        List<Complex>? data = null;
+
+        // We require that the reader be in the right state to read an
+        // object.
+        if (reader.TokenType != JsonTokenType.StartObject)
         {
-            // By the end, we need to have read "v", "dim", and "data".
-            int? version = null;
-            List<int>? dims = null;
-            List<Complex>? data = null;
-
-            // We require that the reader be in the right state to read an
-            // object.
-            if (reader.TokenType != JsonTokenType.StartObject)
+            throw new JsonException();
+        }
+
+        while (reader.Read())
+        {
+            if (reader.TokenType == JsonTokenType.EndObject)
+            {
+                // We're at the end of the array, and can break out of the
+                // read loop.
+                break;
+            }
+
+            // If it's not the end of the object, the current token needs
+            // to be a property name.
+            if (reader.TokenType != JsonTokenType.PropertyName)
             {
                 throw new JsonException();
             }
 
-            while (reader.Read())
+            var propertyName = reader.GetString();
+
+            switch (propertyName)
             {
-                if (reader.TokenType == JsonTokenType.EndObject)
-                {
-                    // We're at the end of the array, and can break out of the
-                    // read loop.
+                case "v":
+                    reader.Read();
+                    version = reader.GetInt32();
                     break;
-                }
 
-                // If it's not the end of the object, the current token needs
-                // to be a property name.
-                if (reader.TokenType != JsonTokenType.PropertyName)
-                {
-                    throw new JsonException();
-                }
+                case "dim":
+                    dims = new List<int>();
+                    reader.Require(JsonTokenType.StartArray);
+                    while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+                    {
+                        dims.Add(reader.GetInt32());
+                    }
+                    break;
```

Comment:
(I can be paranoid here, feel free to ignore)

Is there a chance that at some iteration the `reader.Read()` in line 84 succeeds, but after that the `reader.TokenType` is `null` or some non-integer (and not `JsonTokenType.EndArray` in the end of line 84)? If yes then the subsequent `reader.GetInt32()` in line 86 will [throw](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.utf8jsonreader.getint32?view=net-6.0#exceptions), right?

(I assume that this class is a part of a library with a very broad and unpredictable set of use scenarios, i.e. _any input_ can be fed to this function, including the _malicious input_. If this function is used in an online service, then the malicious input can be used to take the online service down)

Reply:
I do not believe the token type can be a C# `null`, as that indicates the absence of a token (as opposed to `JsonTokenType.Null`). In either case, though, the result from malformed input (whether malicious or due to a bug) should be to throw a JSON serialization exception propagating out to the deserialization call site. In general, using any `System.Text.Json` APIs should be expected to possibly raise that exception such that my understanding would be that when used in a cloud service, checking for that exception would be best practice regardless of this line of code in particular.

Regardless, thanks for raising the question. I don't think it's paranoid at all, but a good discussion nonetheless; I think we should be OK as-is in this case, but I appreciate your raising to help make sure of that!